internal/infrastructure/kubernetes/proxy/resource_provider.go

```diff
@@ -396,15 +423,22 @@ func (r *ResourceRender) PodDisruptionBudget() (*policyv1.PodDisruptionBudget, e
 	}, nil
 }
 
-func (r *ResourceRender) HorizontalPodAutoscaler() (*autoscalingv2.HorizontalPodAutoscaler, error) {
+// HorizontalPodAutoscalerSpec returns the `HorizontalPodAutoscaler` sets spec.
+func (r *ResourceRender) HorizontalPodAutoscalerSpec() (*egv1a1.KubernetesHorizontalPodAutoscalerSpec, error) {
 	provider := r.infra.GetProxyConfig().GetEnvoyProxyProvider()
 	if provider.Type != egv1a1.ProviderTypeKubernetes {
 		return nil, fmt.Errorf("invalid provider type %v for Kubernetes infra manager", provider.Type)
 	}
 
 	hpaConfig := provider.GetEnvoyProxyKubeProvider().EnvoyHpa
+	return hpaConfig, nil
+}
+
+func (r *ResourceRender) HorizontalPodAutoscaler() (*autoscalingv2.HorizontalPodAutoscaler, error) {
+	hpaConfig, err := r.HorizontalPodAutoscalerSpec()
+	// If hpa config is nil, ignore HorizontalPodAutoscaler.
 	if hpaConfig == nil {
-		return nil, nil
+		return nil, err
```

Comment:
Nit: Could we use nil instead of err to tell if a resources is needed? This way improves readability.

Reply:
it is normal logic. look this, L214 to L216, the err will be nil or not nil.

https://github.com/envoyproxy/gateway/blob/a968232b93a0c84b02927c6814ee249cdfa0a761/internal/infrastructure/kubernetes/proxy/resource_provider.go#L209-L221